content/posts/sieve.md

```diff
@@ -9,4 +9,74 @@ author: Bryan Li
+
+**Access Patterns:** Web cache workloads typically follow a power-law (Zipfian) distribution, where a small portion of objects make up the majority of the requests. An important distinction from other cache use cases is that scan and loop/sequential access patterns are rare, and new objects are extremely common, being created every second.
+
+To expand on the power-law distribution, it follows a pattern where the $i^{th}$ most popular object has a relative frequency of $1/i^\alpha$. $\alpha$ here is a parameter determining skewness and has been found to range from 0.55 to 1.5 depending on type of workload, cache level, etc.
+
+**Efficient Cache Eviction Algorithms:** Many of the recent algorithms have greatly increased in complexity, using multiple LRU queues, machine learning, and so on. However, though they might perform better in certain workloads, the improvement is often not significant. In this way the increase in complexity outweighs the benefits.
+
+What makes a cache eviction algorithm more efficient can be boiled down into two parts: **lazy promotion** and **quick demotion**.
+
+- Lazy promotion promotes cached objects only when they are about to be evicted, retaining popular objects with minimal effort.
+  - For comparison, FIFO has no promotion, and LRU promotes on cache hit (eager promotion).
+- Quick demotion removes most objects shortly after they are cached, which takes advantage of the characteristics of the power-law distribution - most objects are unpopular.
+
+## Algorithm
+
+### Overview
+
+SIEVE uses a single FIFO queue and "hand" pointer.
+
+The queue maintains insertion order, and new objects are added to the head of the queue.
+
+Every object in the queue uses 1 bit to track visited/non-visited status, with the hand pointing to the next eviction candidate. The hand pointer moves from the tail to the head.
+
+- **Cache hits** set the visited bit of the cache line to 1.
+  - This also happens to be lock-free, which improves throughput.
+- **Cache misses** look at the head pointer and set the visited bit to 0 if it is 1, moving the hand pointer.
+  - Otherwise (`visited=0`) it will remove the eviction candidate from the cache, and move the hand pointer.
+
+This fulfills the idea of 'lazy promotion' by only promoting at eviction time, by keeping the object at the original location. Because "survived" objects are generally more popular than evicted ones, they will likely be accessed in the future as well. (similar to CLOCK algorithm).
+
+Then, the hand can quickly move towards the head where new objects are inserted, examining new objects shortly after admission. This achieves 'quick demotion'.
+
```

Comment:
Could add this visualization somewhere - https://cachemon.github.io/SIEVE-website/assets/images/illustrations/sieve_diagram.gif